File: thorn-go/internal/scanner/walker.go
package scanner

import (
	"os"
	"path/filepath"
	"unicode/utf8"
)

// maxFileSize skips files larger than this. Secrets/misconfigs live in source and
// config files, never multi-megabyte blobs, so reading huge files whole would waste
// time and memory for no benefit.
const maxFileSize = 5 * 1024 * 1024

var skippedDirs = map[string]bool{".git": true}

type fileContent struct {
	Path    string
	Content string
}

// walkFiles walks root and returns (path, content) for every readable, UTF-8,
// size-bounded file. Unreadable, binary, or oversized files are silently skipped
// rather than failing the whole scan.
func walkFiles(root string) ([]fileContent, error) {
	var files []fileContent

	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			// Can't stat this entry (e.g. permission denied) — skip it, keep walking.
			return nil
		}
		if info.IsDir() {
			if skippedDirs[info.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if info.Size() > maxFileSize {
			return nil
		}

		data, readErr := os.ReadFile(path)
		if readErr != nil {
			return nil
		}
		if !utf8.Valid(data) {
			return nil
		}

		files = append(files, fileContent{Path: path, Content: string(data)})
		return nil
	})

	return files, err
}

File: thorn-rs/src/scanner/walker.rs
use std::fs;
use std::path::{Path, PathBuf};

use walkdir::WalkDir;

/// Files larger than this are skipped — secrets/misconfigs live in source and config files,
/// never in multi-megabyte blobs, and reading huge files whole would waste time and memory.
const MAX_FILE_SIZE_BYTES: u64 = 5 * 1024 * 1024;

const SKIPPED_DIRS: &[&str] = &[".git"];

/// Walks `root` and yields (path, content) for every readable, UTF-8, size-bounded file.
/// Unreadable, binary, or oversized files are silently skipped rather than failing the scan.
pub fn walk_files(root: &Path) -> impl Iterator<Item = (PathBuf, String)> + '_ {
    WalkDir::new(root)
        .into_iter()
        .filter_entry(|entry| {
            if entry.file_type().is_dir() {
                let name = entry.file_name().to_string_lossy();
                return !SKIPPED_DIRS.contains(&name.as_ref());
            }
            true
        })
        .filter_map(|entry| entry.ok())
        .filter(|entry| entry.file_type().is_file())
        .filter_map(|entry| {
            let path = entry.path().to_path_buf();
            let metadata = fs::metadata(&path).ok()?;
            if metadata.len() > MAX_FILE_SIZE_BYTES {
                return None;
            }
            let content = fs::read_to_string(&path).ok()?;
            Some((path, content))
        })
}
